incubator/hnc/pkg/validators/namespace.go

```diff
@@ -0,0 +1,121 @@
+package validators
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/forest"
+)
+
+// NamespaceServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	NamespaceServingPath = "/validate-v1-namespace"
+)
+
+// Note: the validating webhook FAILS CLOSE. This means that if the webhook goes down, all further
+// changes are forbidden.
+//
+// +kubebuilder:webhook:path=/validate-v1-namespace,mutating=false,failurePolicy=fail,groups="",resources=namespaces,verbs=create;update;delete,versions=v1,name=vnamespace.k8s.io
+
+type Namespace struct {
+	Log     logr.Logger
+	Forest  *forest.Forest
+	decoder *admission.Decoder
+}
+
+// nsRequest defines the aspects of the admission.Request that we care about.
+type nsRequest struct {
+	ns *corev1.Namespace
+	op v1beta1.Operation
+}
+
+// Handle implements the validation webhook.
+func (v *Namespace) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := v.Log.WithValues("NamespaceName", req.Name)
+	// Early exit since the HNC SA can do whatever it wants.
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		log.V(1).Info("Allowed change by HNC SA")
+		return allow("HNC SA")
+	}
+
+	decoded, err := v.decodeRequest(log, req)
+	if err != nil {
+		log.Error(err, "Couldn't decode request")
+		return deny(metav1.StatusReasonBadRequest, err.Error())
+	}
+
+	resp := v.handle(decoded)
+	log.V(1).Info("Handled", "allowed", resp.Allowed, "code", resp.Result.Code, "reason", resp.Result.Reason, "message", resp.Result.Message)
+	return resp
+}
+
+// handle implements the non-boilerplate logic of this validator, allowing it to be more easily unit
+// tested (ie without constructing a full admission.Request).
+func (v *Namespace) handle(req *nsRequest) admission.Response {
+	owner := req.ns.Annotations[api.AnnotationOwner]
+	if owner != "" && req.op == v1beta1.Delete {
+		msg := fmt.Sprintf("The namespace %s is an owned namespace. Please delete the HNS instance from the owner namespace %s instead.", req.ns.Name, owner)
+		return deny(metav1.StatusReasonForbidden, msg)
+	}
+
+	return v.checkForest(req)
+}
+
+func (v *Namespace) checkForest(req *nsRequest) admission.Response {
+	v.Forest.Lock()
+	defer v.Forest.Unlock()
+
+	ns := v.Forest.Get(req.ns.Name)
+
+	// Check whether the deleting namespace has owned namespaces if it doesn't allow cascading deletion.
+	if req.op == v1beta1.Delete && !ns.AllowsCascadingDelete() {
+		hnses := []string{}
+		for _, cnm := range ns.ChildNames() {
+			cns := v.Forest.Get(cnm)
+			if cns.IsOwned {
+				hnses = append(hnses, cnm)
+			}
```

Comment:
Shouldn't this be `if cns.IsOwned && !cns.AllowsCascadingDelete()`? If so, maybe change `hnses` to `cantDelete` or something like that, to make it clear why things get added to that list.

If I'm correct, and this is a bug, maybe add a unit test for this too?

Reply:
Yes good catch it's a bug. Say the owner doesn't allowCD but all its owned NS allows, then it should return `allow()`. Good catch on this corner case, will add the unit test on that and manual test it too.

Reply:
Added new unit tests and manually tested with setting `allowCascadingDelete`s (the webhook and cascading deletion all worked as expected).